### ISSUE TITLE
pathutils.relativeTo is buggy on Windows

### ISSUE THREAD

**Issue by skilchen** [nim-lang/Nim#9612]
### Example
```nim
import pathutils

proc testDifferentDrives() =
  var absFile = AbsoluteFile(r"c:\a\b\c")
  var absDir = AbsoluteDir(r"d:\a\b")
  
  echo string(relativeTo(absFile, absDir)), " is not a valid relative path"
  
proc testDrive_NoDrive() =
  var absFile = AbsoluteFile(r"c:\a\b\c")
  var absDir = AbsoluteDir(r"\a\b\d")
  echo string(relativeTo(absFile, absDir)), " is not a valid relative path, should be: \"..\\c\""
  absFile = AbsoluteFile(r"\a\b\c")
  absDir = AbsoluteDir(r"c:\a\b\d")
  echo string(relativeTo(absFile, absDir)), " is an ugly relative path, should be: \"..\\c\""
  
proc testSameDirAsFile() =
  var p = r"c:\a\b\c"
  echo string(relativeTo(AbsoluteFile(p), AbsoluteDir(p))), " but produces an empty path"
  
proc testMixedSeparators() =
  var absFile = AbsoluteFile("/a/b/c")
  var absDir = AbsoluteDir(r"\a\b")
  echo string(relativeTo(absFile, absDir)), " is an ugly relative path, should be: \"c\""
  
echo "relativeTo on different drives should raise exception"
testDifferentDrives()
echo ""
echo "if only one path has a drive, the drive should be ignored"
testDrive_NoDrive()
echo ""
echo "mixed dirSeps should be normalized"
testMixedSeparators()
echo ""
echo "if file and basedir are the same, should either raise exception or return \".\""
testSameDirAsFile()
```

### Current Output
```
relativeTo on different drives should raise exception
..\..\..\c:\a\b\c is not a valid relative path

if only one path has a drive, the drive should be ignored
..\..\..\..\c:\a\b\c is not a valid relative path, should be: "..\c"
..\..\..\..\\a\b\c is an ugly relative path, should be: "..\c"

mixed dirSeps should be normalized
..\..\..\/\a\b\c is an ugly relative path, should be: "c"

if file and basedir are the same, should either raise exception or return "."
 but produces an empty path
```

**Comment by dom96** [comment]
``pathutils``? The compiler module?

**Comment by skilchen** [comment]
yes

**Comment by skilchen** [comment]
@nc-x what i expect from relativeTo above is based on what Python's os.path.relpath produces:
In Python:
- there is no relative path across drive boundaries:
  ```python
  os.path.relpath(r"c:\a\b\c", r"d:\a\b")
   ```
   raises `ValueError: path is on drive c:, start on drive d:`

- if one path has a drive letter and the other not, the drive letter is ignored 
  ```python
   os.path.relpath(r"c:\a\b\c", r"\a\b\d")
   os.path.relpath(r"\a\b\c", r"c:\a\b\d")
   ```
   both produce `..\c`

- if one path uses `\` as dirSep and the other `/`:
  ```python
   os.path.relpath(r"/a/b/c", r"\a\b")
   ```
   produces: `c`

- if `filename` and `basedir` are the same:
  ```python
   os.path.relpath(r"c:\a\b\c", r"c:\a\b\c")
   ```
   produces: `.`
   this is probably not what Nim wants to do because the distinction between `AbsoluteFile` and `AbsoluteDir` implies that a path can not simultaneously be an `AbsoluteFile` and an `AbsoluteDir`.

It is questionable if mixing "path" and "dir" and "file" is a good design. Conceptually, a "path" is just an entry in a directory table that does not imply whether it is a file or a directory or a device or a named pipe or a socket, or whatever else might have an entry in a directory table.

**Comment by timotheecour** [comment]
> this is probably not what Nim wants to do because the distinction between AbsoluteFile and AbsoluteDir implies that a path can not simultaneously be an AbsoluteFile and an AbsoluteDir.

that's one of the issues with `pathutils` that I'm raising in https://github.com/nim-lang/Nim/issues/9508

**Comment by skilchen** [comment]
I don't understand exactly what you are asking me. 

IMHO a path handling module should not care about what a path points to. A `relativeTo` in the stdlib should take two paths as input and return a path or raise an exception if one path can not be made relative to the other. A relative path is only useful if it actually saves some of the expensive directory reads  needed to find the thing the relative path points to in the filesystem. Relative paths going up to the root and then down to the target are useless or even harmful.

IMHO, it would be a good idea to just reimplement what Python does in `os.paths`.  

In this issue my only concern is, that on Windows `pathutils.relativeTo` sometimes returns invalid `RelativeFile`s. Thats what some people have observed via strange looking error messages.

**Comment by demotomohiro** [comment]
> proc testDifferentDrives() =
>   var absFile = AbsoluteFile(r"c:\a\b\c")
>   var absDir = AbsoluteDir(r"d:\a\b")
>   
>   echo string(relativeTo(absFile, absDir)), " is not a valid relative path"

This case should raise exception. Returning first argument as is is not a option as return value must be relative path.
``relativeTo`` needs to check drive of input paths and raise exception if they don't have same drive letter.

> proc testDrive_NoDrive() =
>   var absFile = AbsoluteFile(r"c:\a\b\c")
>   var absDir = AbsoluteDir(r"\a\b\d")
>   echo string(relativeTo(absFile, absDir)), " is not a valid relative path, should be: \"..\\c\""
>   absFile = AbsoluteFile(r"\a\b\c")
>   absDir = AbsoluteDir(r"c:\a\b\d")
>   echo string(relativeTo(absFile, absDir)), " is an ugly relative path, should be: \"..\\c\""

I wonder these paths without drive letter are really absolute path.
I think a absolute path should be able to identify a file or directory without any additional information.
These drive letterless paths are different from relative path, but which file/directory the path point to depend on current drive.
But Microsoft says it is an absolute path if it begin with A single backslash:
https://docs.microsoft.com/windows/desktop/FileIO/naming-a-file#fully-qualified-vs-relative-paths
Whether it is absolute path and whether it can be a AbsoluteFile or AbsoluteDir type value is a different problem?
I'm not sure this is a good idea but forbiding storing absolute path without drive letter to ``AbsoluteFile``/``AbsoluteDir`` would keep pathutils simple.
In that case, ``toAbsolute`` proc add drive letter if ``file`` is such a path.

> proc testMixedSeparators() =
>   var absFile = AbsoluteFile("/a/b/c")
>   var absDir = AbsoluteDir(r"\a\b")
>   echo string(relativeTo(absFile, absDir)), " is an ugly relative path, should be: \"c\""

It seems current ``relativeTo`` works paths with mixed path separater, but doesn't support absolute path without drive letter.
When a first charactor of a path is \ or /, ``proc next(it: var PathIter; x: string): (int, int) =`` returns the first charactor as is.
And path comparsion in ``relativeTo`` treat \ and / are different.

```nim
let absX = toAbsolute("c:/a/b/c", AbsoluteDir"c:\")
let absY = toAbsolute(r"c:\a\b", AbsoluteDir"c:")
echo relativeTo(absX, absY.AbsoluteDir).string
```

Output:
```
c
```

I heard that compiler/pathutil replace ospaths module but it still not ready to take general windows paths.

I also think ``relativeTo`` should normalize input path (or should I say canonicalize as ``canon`` proc make a cannonical path in ``pathutil``?) if ``AbsoluteFile`` or ``AbsoluteDir`` are not always canonical path.

```nim
let absX = toAbsolute(r"c:\a\b\c\..\c", AbsoluteDir"c:\")
let absY = toAbsolute(r"c:\a\b", AbsoluteDir"c:\")
echo absX.string
echo absY.string
echo relativeTo(absX, absY.AbsoluteDir).string
```
output:
```
c:\a\b\c\..\c
c:\a\b
c\..\c
```

> A relative path is only useful if it actually saves some of the expensive directory reads needed to find the thing the relative path points to in the filesystem.

Current ``relativeTo`` implemention doesn't access filesystem.

> IMHO, it would be a good idea to just reimplement what Python does in `os.paths`.

``relativeTo`` in compiler/pathutil module takes only absolute path but ``os.paths`` in Python can take relative path.
And a function that calculate a relative path from relative paths are complicated thing.
I tried to make a such procedure before but I gave up as pathutil already have one. And I dont know whether there are people who needs to convert a relative path X relative to Y to a relative path X' relative to Z.
https://github.com/nim-lang/Nim/pull/8166
